src/camera.rs

```diff
@@ -27,92 +34,151 @@ pub struct Camera {
         }
 
-        // Collision detection (xz-plane)
-        let is_free = world.is_position_free(&next_pos_amplified);
+        displacement
+    }
+
+    pub fn step(&mut self, elapsed: Duration, world: &World) {
+        // Compute the next position
+        let dt = elapsed.as_secs_f32();
 
-        // Free-fall handling
-        let is_falling = world.is_position_free_falling(&next_pos_amplified);
-        let dz_fall = self.gravity_handler.step(is_falling, elapsed);
-        next_pos[1] -= dz_fall;
+        // add gravity
+        if self.in_air {
+            self.velocity += GRAVITY_ACCELERATION_VECTOR * dt;
+        }
+
+        // TODO will have to do something cleaner when other sources of
+        //      horizontal velocity will be implemented
+        {
+            let controls_vel = self.controls_velocity();
+            self.velocity[0] = controls_vel[0];
+            self.velocity[2] = controls_vel[2];
+        }
 
-        // Position update
-        if is_free {
-            self.position.set_position(next_pos);
-        } else { 
-            
+        let mut dt = elapsed.as_secs_f32();
+        dt = dt - self.move_with_collision(dt, world);
+        if dt > 0. {
+            dt = dt - self.move_with_collision(dt, world);
         }
+        // TODO we may want to do it a third time, to handle x-y-z diagonal
+        //      movements. However, there are no slopes in the cubes for now, so it
+        //      is not necessary
+
+        // update in_air
+        let displacement = Vector3::new(0., -1e-5, 0.);
+        self.in_air = !world.collides(&Self::make_aabb(&(&self.position + displacement)));
 
         self.compute_selected_cube(world);
     }
 
+    /// Integrate the velocity to move the camera, with collision. Returns the
+    /// dt (in seconds), which can be smaller than `dt` if there is a collision.
+    fn move_with_collision(&mut self, dt: f32, world: &World) -> f32 {
+        let target = Self::make_aabb(&(&self.position + self.velocity * dt));
+
+        let (collision_time, normal) =
+            world.collision_time(&Self::make_aabb(&self.position), &target, &self.velocity);
+
+        if collision_time >= dt {
+            // can move straight away
+            self.position += self.velocity * dt;
+
+            dt
+        } else {
+            // we want to put a margin, to avoid collision even with floats rounding
+
+            // TODO if we are here, we can assume the velocity is nonzero I
+            // think, but I am not sure
+            let dtmargin = 1e-5 / self.velocity.norm();
+            self.position += self.velocity * (collision_time - dtmargin);
+
+            // remove component of velocity along the normal
+            let vnormal = normal * normal.dot(&self.velocity);
+            self.velocity = self.velocity - vnormal;
+
+            collision_time
+        }
+    }
+
+    fn make_aabb(position: &Position) -> AABB {
+        let diameter = 0.5;
+        let forehead = 0.1;
+
```

Comment:
You can mark them as `const` instead of `let` (and you must declare their types)